### PR TITLE
MicroBit Detect Board and Buttons A/B (2)

### PR DESCRIPTION
**Pull request by epeach**
# Description
Woohoo! Getting Microbit into the AppLab. Rough MVP to get to demo.

Example App using MicroBit:
ThumbWars

## Links

Requires a MB board flashed with MB firmata.

Behind experiment 'microbit'.

## Testing story
Current plan is to write tests after the demo to keep forward momentum as much as possible.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked
cc: @joshlory

### CODE REVIEW

File: apps/src/lib/kits/maker/Button.js [hunk 1]
```diff
@@ -19,4 +19,25 @@ export default function Button(opts) {
     get: () => this.value === 1
   });
 }
-Button.inherits(five.Button);
+PlaygroundButton.inherits(five.Button);
+
+export function MicroBitButton(board) {
+  // There are six button events, ['', 'down', 'up', 'click', 'long-click', 'hold']
```
Comment (ajpal) [inline apps/src/lib/kits/maker/Button.js:25]: What does `''` mean as a button event?

Comment (epeach) [reply apps/src/lib/kits/maker/Button.js:25]: Unknown. I referenced this from the MicroBit firmata tests -> https://github.com/microbit-foundation/microbit-firmata/blob/fd572115cf69a820d0d814736f8572243f19af7a/client/mbTests.js#L325
and having ' ' makes the rest of the events align.

File: apps/src/lib/kits/maker/Button.js [hunk 2]
```diff
@@ -19,4 +19,25 @@ export default function Button(opts) {
...
+  this.buttonEvents = new Array(6).fill(0);
+  board.mb.addFirmataEventListener((sourceID, eventID) => {
+    if (board.pin === sourceID) {
+      this.buttonEvents[eventID]++;
```
Comment (ajpal) [inline apps/src/lib/kits/maker/Button.js:29]: What does this do? When is it decremented?

File: apps/src/lib/kits/maker/Button.js [hunk 3]
```diff
@@ -19,4 +19,25 @@ export default function Button(opts) {
...
+        this.emit('down');
+      } else if (eventID === 2) {
+        this.emit('up');
+      }
+    }
+  });
+
+  // Add a read-only `isPressed` property
+  Object.defineProperty(this, 'isPressed', {
+    get: () => this.buttonEvents[1] > 0
```
Comment (ajpal) [inline apps/src/lib/kits/maker/Button.js]: Why index 1?

File: apps/src/lib/kits/maker/MicroBitBoard.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
+/** @file Board controller for BBC micro:bit */
+/* global SerialPort */ // Maybe provided by the Code.org Browser
```
Comment (ajpal) [inline apps/src/lib/kits/maker/MicroBitBoard.js:2]: Why maybe?

File: apps/src/lib/kits/maker/Button.js (hunk 2 above)
Comment (epeach) [reply apps/src/lib/kits/maker/Button.js:29]: Each type of button event is initialized at zero, so this counts the number of times an event has occurred. Whether this needs to be constantly counted upwards or whether we could just increment from zero to one (i.e. not yet pressed to has been pressed), is something I don't have a strong preference on.

File: apps/src/lib/kits/maker/Button.js (hunk 3 above)
Comment (epeach) [reply apps/src/lib/kits/maker/Button.js]: Index 1 is the 'button went down' event.

File: apps/src/lib/kits/maker/MicroBitBoard.js (hunk 4 above)
Comment (epeach) [reply apps/src/lib/kits/maker/MicroBitBoard.js:2]: If this code is being run in the browser, this value is provided by https://github.com/code-dot-org/browser/blob/a9bea833650028f4a34f211d02b12dbe042959b2/src/preload.js#L23

File: apps/src/lib/kits/maker/Button.js (hunk 3 above)
Comment (ajpal) [reply apps/src/lib/kits/maker/Button.js]: Sorry might be missing something obvious but: Wouldn't this return true always after the first time the button is pressed, not just when the button is *currently* pressed?

Comment (epeach) [reply apps/src/lib/kits/maker/Button.js]: Good catch - updated!

**Review by ajpal — APPROVED** [review]

**Pull request merged**